packages/flutter/lib/src/material/colors.dart

```diff
@@ -88,6 +90,133 @@ class MaterialAccentColor extends ColorSwatch<int> {
   Color get shade700 => this[700];
 }
 
+/// Signature for the function that returns a (potentially different) color
+/// based on a given set of states.
+typedef ColorResolver = Color Function(Set<MaterialState> states);
+
+/// Defines a [Color] that can react to changes in the state of a Material
+/// component, based on a given set of [MaterialState]s.
+///
+/// This is useful for preserving the accessibility of text in different states
+/// of a component. For example, in a [FlatButton] with blue text, the text will
+/// become more difficult to read when the button is hovered, focused, or pressed,
+/// because the contrast ratio between the button and the text will decrease. To
+/// solve this, you can use [MaterialStateColor] to make the text darker when the
+/// [FlatButton] is hovered, focused, or pressed.
+///
+/// To use a [MaterialStateColor], you can either:
+///   1. Create a subclass of [MaterialStateColor] and implement the abstract `resolve` method.
+///   2. Use [MaterialStateColor.resolveWith] and pass in a callback that
+///      will be used to resolve the color in the given states.
+///
+/// This should only be used as parameters when they are documented to take
+/// [MaterialStatefulColor], otherwise only the default state will be used.
+///
+/// Example:
```

Comment:
This should probably be an `@tool sample`

Reply:
Done.